### PR TITLE
Guard product page when product is null

### LINKED ISSUE #41292: Invalid product URLs trigger Fatal in ProductController::assignPriceAndTax (PHP 8 / PS 9)
### Prerequisites

- [x] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [x] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue%20type%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

`ProductController::initContent()` unconditionally calls `assignPriceAndTax()` without checking whether `$this->product` is a loaded object. When the product cannot be loaded (deleted, inactive, invalid id in the URL), `init()` already
 handled the case by setting the 404 template and returning , but `Controller::run()` keeps going and calls `initContent()` anyway, leading to a Fatal error.

**Call chain:**
```
Controller::run() [classes/controller/Controller.php:308]
  → ProductController::initContent() [controllers/front/ProductController.php:306]
    → ProductController::assignPriceAndTax() [line 529]
       → crash on line 536 (getTaxesRate on null) OR line 539 (getPriceStatic with null id)
```

**Two observable crash points:**

- `ProductController.php:536` , `$this->product->getTaxesRate(...)` → Fatal `Call to a member function getTaxesRate() on null` when `$this->product === null` (URL has `id_product=0` or missing id).
- `ProductController.php:539` , `Product::getPriceStatic($this->product->id, ...)` → `PrestaShopException('Product ID is invalid.')` when `$this->product` is a non-loaded Product object (`new Product(XXXXX)` where the id doesn't exist
returns an object with `id = null`).

**Why `init()` doesn't prevent it:** `init()` correctly detects the invalid product (line 133) and calls `setTemplate('errors/404'); return;`, but the `return` only exits `init()`, not `run()`. `Controller::run()` proceeds to call
`initContent()` regardless.

**Why this is a PHP 8 regression:** PS 1.7 / PS 8 `ProductController` had the same logical flaw, but PHP 7 tolerated `null->method()` with a silent warning. PHP 8 turns it into a `TypeError`.

**Real-world impact:** on any production store with indexed URLs, search engine crawlers (Google, Bing, Facebook) continuously hit URLs of deleted/renamed products. On our store (540k orders, 50k+ products, multilingual) this produces
~40 `*ERROR* Product ID is invalid` entries per hour in `var/logs/<YYYYMMDD>_exception.log`, plus occasional 500s in Apache error log. Zero functional impact (caught exceptions are silenced upstream for most code paths), but the log
noise drowns out legitimate errors.

### Proposed fix options

1. **Guard at the top of `assignPriceAndTax()`:**
   ```php
   protected function assignPriceAndTax(): void
   {
       if (!Validate::isLoadedObject($this->product)) {
           return;
       }
       // ... existing code ...
   }
   ```
2. **Guard at the top of `initContent()`** , protects all `assign*` helpers (`assignCategory`, `assignAttributesCombinations`, etc.) which have the same latent issue.
3. **Abort the run flow in `Controller::run()`** when a 404 template was already set by `init()`. Most invasive, but fixes the class of bugs globally (`CategoryController::getCategoryToRedirectTo()` has the same pattern, crashing on
orphaned inactive categories).

Option 2 is the cleanest local fix.

### Steps to reproduce

1. On any PS 9 install, hit a URL of a non-existent or deleted product, e.g.:
   ```
   https://<shop>/<lang>/999999-nonexistent.html
   https://<shop>/<lang>/<id_of_deleted_product>-<old-slug>.html
   ```
2. Browser receives `HTTP 500`.
3. Check `var/logs/<YYYYMMDD>_exception.log` and `/var/log/apache2/error.log`:
   - `PHP Fatal error: Uncaught Error: Call to a member function getTaxesRate() on null in controllers/front/ProductController.php:536`
   - And/or `*ERROR* Product ID is invalid. at line 3363 in file classes/Product.php`
4. Stack trace consistently terminates at `ProductController::assignPriceAndTax()` → `initContent()` → `Controller::run()`.

### Expected behavior

A request to a non-existent or deleted product URL should return HTTP 404 rendered via the `errors/404` template that `init()` already selected, with no Fatal error, no PrestaShopException, and no entry in `exception.log`.

### Actual Result

HTTP 500 (or silent caught exception), every hit produces at least one log entry. On a shop with crawler traffic, `exception.log` accumulates thousands per day.

### PrestaShop version where the bug happens

9.1.0, 8.2.5

### How have you installed PrestaShop

_No response_

### PHP version(s) where the bug happened

8.1

### Your company or customer's name goes here (if applicable).

_No response_

### PR DESCRIPTION
**Pull request by Codencode**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.1.x
| Description?      | This PR adds a null guard around product-dependent logic in `ProductController` to avoid errors when the product is not available or not properly initialized.<br/>It ensures methods like `initContent()` and breadcrumb generation do not try to access product data when `$this->product` is `null`, making the product page flow more robust in invalid or edge-case contexts.
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | In the front office, it tries to access a non-existent article by entering a non-existent product ID.
| UI Tests          | 🟢 https://github.com/Codencode/ga.tests.ui.pr/actions/runs/24670317307
| Fixed issue or discussion?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/41292 https://github.com/PrestaShop/PrestaShop/discussions/36868 #41517 
| Related PRs       | 
| Sponsor company   | Codencode snc

### CODE REVIEW

**Comment by Codencode** [comment]
@jf-viguier,

I have recreated this new PR from scratch, but the result is the same. Maybe we are doing something wrong?
This PR was created starting from 9.1.x; you can verify it here:
https://github.com/Codencode/PrestaShop/blob/a0829c086260799bd6e99777d5271427a2263238/src/Core/Version.php#L15

**Review by Hlavtox — DISMISSED** [review]
So this is a replacement for the errors condition we removed. :-)

It's OK, but I would prefer the controller just killing itself and redirecting to some 404 logic.

**Comment by Codencode** [comment]
> So this is a replacement for the errors condition we removed. :-)
> 
> It's OK, but I would prefer the controller just killing itself and redirecting to some 404 logic.

So, would you prefer that the system performs a redirect to the 404 controller?

**Comment by Hlavtox** [comment]
@Codencode Nono, in my opinion, if a URL throws 404, the URL should stay.

What I meant is something like:
if product doesnt exist
=> return new NotFoundController()

**Comment by Codencode** [comment]
@Hlavtox 
Ok, now it's clear.
I would probably leave it as it is for now, and consider it in a separate PR if needed.

**Comment by Codencode** [comment]
@PrestaShop/committers ping for review

File: controllers/front/ProductController.php [hunk 1]
```diff
@@ -296,102 +297,104 @@ public function initializeCategory(): void
      */
     public function initContent(): void
     {
-        // Assign template vars related to the category + execute hooks related to the category
-        $this->assignCategory();
+        if ($this->product !== null) {
```
Comment (cnavarro-prestashop) [inline controllers/front/ProductController.php]: Instead wrapping all the code in a condition, why not return making an early return and calling initContent ?

```suggestion
        if ($this->product === null) {
            parent::initContent();
            return;
        }
```
If the product is not null, it will execute all the code then initContent method as planned.

File: controllers/front/ProductController.php [hunk 2]
```diff
@@ -1379,38 +1382,40 @@ public function getBreadcrumbLinks(): array
...
-        // $productBreadcrumbCategory can have two possible values
-        // - current : Category the product was accessed from
-        // - default : Product default category
-        if (('current' === Configuration::get('PS_PRODUCT_BREADCRUMB_CATEGORY'))
-            && !empty($this->category)) {
-            $productBreadcrumbCategory = $this->category;
-        } else {
-            $productBreadcrumbCategory = new Category($this->product->id_category_default, $this->context->language->id);
-        }
+        if ($this->product !== null) {
```
Comment (cnavarro-prestashop) [inline controllers/front/ProductController.php]: Same comment / logic here : 

```suggestion
        if ($this->product === null) {
            return $breadcrumb;
        }
```

**Review by cnavarro-prestashop — COMMENTED** [review]
Only 2 comments but I agree with @Hlavtox about the solution to redirect to PageNotFoundController, it would be prettier as fix.

**Comment by Codencode** [comment]
@cnavarro-prestashop

I followed your advice about exiting immediately if the product doesn't exist, so I made the change.

Regarding returning PageNotFoundController directly, I'm not sure how it should be implemented correctly.

Would it be acceptable to replace:
https://github.com/PrestaShop/PrestaShop/blob/ba29cf3e874db27b8a03ffebc0f9b47cb5aee876/controllers/front/ProductController.php#L133-L140

with:

```php
        if (!Validate::isLoadedObject($this->product)) {
            /** @var PageNotFoundController $controller */
            $controller = Controller::getController('PageNotFoundController');
            self::$initialized = false;
            $controller->init();
            $controller->run();

            exit;
        }
```

I can't think of anything else.

**Comment by Hlavtox** [comment]
@Codencode @cnavarro-prestashop I opened an issue about this some time ago - https://github.com/PrestaShop/PrestaShop/discussions/36868.

Too bad we can't do something like Symfony's `NotFoundException` and call it a day.

Now we have different approaches:
- Some pages set a 404 error template, like product page, but you have to do several conditions, some code can still run and fail.
- Some pages redirect to 404, which is wrong, because the signal to search engines is not proper. The URL must stay, with a 404 response.

**Review by Hlavtox — APPROVED** [review]

**Review by mattgoud — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]

**Comment by jolelievre** [comment]
@Codencode don't use `exit` in the code if you can avoid it; it has really bad side effects, like breaking/stopping the phpunit tests, and should be banned from the codebase

Although, because of the legacy architecture, sometimes it's, sadly, the only solution, but only as a last resort

**Comment by Codencode** [comment]
@jolelievre  
Thank you for the explanation, but in the example code I posted, we need to interrupt the code flow because we are forcing the execution of another controller.  
Can this be done in another way?

**Review by AureRita — APPROVED** [review]
Hi @Codencode 

Thank you for your PR, I tested it and it seems to works as you can see : 

https://github.com/user-attachments/assets/596e1bb5-50b3-48c7-8f22-689860055980

Because the PR seems to works as expected, It's QA :heavy_check_mark: 

Thank you
